signer_client.go

```diff
@@ -215,26 +225,70 @@ func (s *signerClient) RawClientWithMacAuth(
 	return s.signerMac.WithMacaroonAuth(parentCtx), s.timeout, s.client
 }
 
-func marshallSignDescriptors(
-	signDescriptors []*SignDescriptor) []*signrpc.SignDescriptor {
+func marshallSignDescriptors(signDescriptors []*SignDescriptor,
+	fullDescriptors bool) []*signrpc.SignDescriptor {
 
-	rpcSignDescs := make([]*signrpc.SignDescriptor, len(signDescriptors))
-	for i, signDesc := range signDescriptors {
-		var keyBytes []byte
-		var keyLocator *signrpc.KeyLocator
-		if signDesc.KeyDesc.PubKey != nil {
-			keyBytes = signDesc.KeyDesc.PubKey.SerializeCompressed()
+	// partialDescriptor is a helper method that creates a partially
+	// populated sign descriptor that is backward compatible with the way
+	// some applications like Loop expect the call to lnd to be made. This
+	// function only populates _either_ the public key or the key locator in
+	// the descriptor, but not both.
+	partialDescriptor := func(
+		d keychain.KeyDescriptor) *signrpc.KeyDescriptor {
+
+		keyDesc := &signrpc.KeyDescriptor{}
+		if d.PubKey != nil {
+			keyDesc.RawKeyBytes = d.PubKey.SerializeCompressed()
 		} else {
-			keyLocator = &signrpc.KeyLocator{
-				KeyFamily: int32(
-					signDesc.KeyDesc.KeyLocator.Family,
-				),
-				KeyIndex: int32(
-					signDesc.KeyDesc.KeyLocator.Index,
-				),
+			keyDesc.KeyLoc = &signrpc.KeyLocator{
+				KeyFamily: int32(d.KeyLocator.Family),
+				KeyIndex:  int32(d.KeyLocator.Index),
 			}
 		}
 
+		return keyDesc
+	}
+
+	// fullDescriptor is a helper method that creates a fully populated sign
+	// descriptor that includes both the public key and the key locator (if
+	// available). For the locator we explicitly check that both the family
+	// _and_ the index is non-zero. In some applications it's possible that
+	// the family is always set (because only a specific family is used),
+	// but the index might be zero because it's the first key, or because it
+	// isn't known at that particular moment.
+	// We aim to be compatible with this method in lnd's wallet:
+	// https://github.com/lightningnetwork/lnd/blob/master/lnwallet/btcwallet/signer.go#L286
+	// Because we know all custom families (0 to 255) are derived at wallet
+	// creation, and the very first index of each family/account is always
+	// derived, we know that only using the public key for that very first
+	// index will work. But for a freshly initialized wallet (e.g. restored
+	// from seed), we won't know any indexes greater than 0, so we _need_ to
+	// also specify the key locator and not just the public key.
+	fullDescriptor := func(
```

Comment:
I wanted to make this as clear and obvious as possible, with as much comment on _why_ we're doing it that way, so I went with two closures on purpose. Just so nobody else spends debug hours on this (because this is already the second or third time I ran into this, once in Loop and once in Pool).